modules/renter/uploadheap.go

```diff
@@ -236,6 +238,30 @@ func (r *Renter) managedBuildChunkHeap(hosts map[string]struct{}) {
 	// the heap.
 	id := r.mu.Lock()
 	for _, file := range r.files {
+		// check for local file
+		tf, exists := r.persist.Tracking[file.name]
+		if exists {
+			// Build maps that map contract id to its offline and goodForRenew
+			// status
+			contractIDs := make(map[types.FileContractID]struct{})
+			for cid := range file.contracts {
+				contractIDs[cid] = struct{}{}
+			}
+			goodForRenew := make(map[types.FileContractID]bool)
+			offline := make(map[types.FileContractID]bool)
+			for cid := range contractIDs {
+				resolvedID := r.hostContractor.ResolveID(cid)
+				cu, ok := r.hostContractor.ContractUtility(resolvedID)
+				goodForRenew[cid] = ok && cu.GoodForRenew
+				offline[cid] = r.hostContractor.IsOffline(resolvedID)
+			}
+			// Check if local file is missing and redundancy is less than 1
+			// log warning to renter log
+			if _, err := os.Stat(tf.RepairPath); os.IsNotExist(err) && file.redundancy(offline, goodForRenew) < 1 {
+				r.log.Println("File not found on disk:", tf.RepairPath)
+			}
+		}
```

Comment:
is it possible to move the `offline` and `goodForRenew` map construction outside the loop? I think that's what `FileList` does. It's more efficient and it makes the check much more readable:
```go
// Check for "endangered" files. A file is endangered if its redundancy is
// less than 1 and no local copy is available.
if tf, exists := r.persist.Tracking[file.name]; exists {
    if _, err := os.Stat(tf.RepairPath); os.IsNotExist(err) && file.redundancy(offline, goodForRenew) < 1 {
        r.log.Println("File not found on disk:", tf.RepairPath)
    }
}

Reply:
`FileList` loops over the files to get a slice of `files` and `contractIDs` then loops over the `contractsIDs` to make the maps of `offline` and `goodForRenew`, and finally loops over the slice of `files` to generate the final splice of `filelist`.

So I could move the map contruction to its own for loop to help with the readability, but would it really make it that much more efficient?  In both cases you need to have the nested for loops of the getting the `contractIDs`, then its another for loop to generate the maps and one last one for check the file redundancy.  Isn't it O(nm) either way?

Reply:
@lukechampine I don't think we need to move it outside the loop but we can get rid of the `if exists` by doing something like the following:
```
// check for local file
tf, exists := r.persist.Tracking[file.name]
if !exists {
    r.log.Critical("Some message about why that shouldn't happen")
}

// Build maps and compute redundancy

// Check if local file is missing and redundancy is less than 1
// log warning to renter log
if _, err := os.Stat(tf.RepairPath); (!exists || os.IsNotExist(err)) && file.redundancy(offline, goodForRenew) < 1 {
    r.log.Println("File not found on disk:", tf.RepairPath)
```

Reply:
my understanding was that there would be some overlap between files in the `offline` and `goodForRenew` maps, since files share contracts. So if you move the construction outside the loop, you only calculate the status of each contract once, whereas if it's inside the loop, you end up recalculating the status of each contract multiple times.

Reply:
So would this be what you were thinking @lukechampine 
```golang
func (r *Renter) managedBuildChunkHeap(hosts map[string]struct{}) {
	// Loop through the whole set of files and get a list of chunks to add to
	// the heap.
	id := r.mu.RLock()
	goodForRenew := make(map[types.FileContractID]bool)
	offline := make(map[types.FileContractID]bool)
	for _, file := range r.files {
		file.mu.RLock()
		// Build maps that map contract id to its offline and goodForRenew
		// status
		contractIDs := make(map[types.FileContractID]struct{})
		for cid := range file.contracts {
			contractIDs[cid] = struct{}{}
		}
		for cid := range contractIDs {
			resolvedID := r.hostContractor.ResolveIDToPubKey(cid)
			cu, ok := r.hostContractor.ContractUtility(resolvedID)
			goodForRenew[cid] = ok && cu.GoodForRenew
			offline[cid] = r.hostContractor.IsOffline(resolvedID)
		}

		unfinishedUploadChunks := r.buildUnfinishedChunks(file, hosts)
		for i := 0; i < len(unfinishedUploadChunks); i++ {
			r.uploadHeap.managedPush(unfinishedUploadChunks[i])
		}
		file.mu.RUnlock()
	}
	for _, file := range r.files {
		file.mu.RLock()
		// check for local file
		tf, exists := r.persist.Tracking[file.name]
		if exists {
			// Check if local file is missing and redundancy is less than 1
			// log warning to renter log
			if _, err := os.Stat(tf.RepairPath); os.IsNotExist(err) && file.redundancy(offline, goodForRenew) < 1 {
				r.log.Println("File not found on disk and possibly unrecoverable:", tf.RepairPath)
			}
		}
		file.mu.RUnlock()
	}
	r.mu.RUnlock(id)
}
```